x/posts/keeper/msg_server.go

```diff
@@ -498,3 +498,92 @@ func (m msgServer) UpdateParams(goCtx context.Context, msg *types.MsgUpdateParam
 
 	return &types.MsgUpdateParamsResponse{}, nil
 }
+
+// MovePost defines the rpc method for Msg/MovePost
+func (k msgServer) MovePost(goCtx context.Context, msg *types.MsgMovePost) (*types.MsgMovePostResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Make sure the post exists
+	post, found := k.GetPost(ctx, msg.SubspaceID, msg.PostID)
+	if !found {
+		return nil, errors.Wrapf(sdkerrors.ErrInvalidRequest, "post with id %d does not exist", msg.PostID)
+	}
+
+	// Check if the target subspace exists
+	if !k.HasSubspace(ctx, msg.TargetSubspaceID) {
+		return nil, errors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace with id %d not found", msg.TargetSubspaceID)
+	}
+
+	// Check if the target section exists
+	if !k.HasSection(ctx, msg.TargetSubspaceID, msg.TargetSectionID) {
+		return nil, errors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace section with id %d not found", msg.TargetSectionID)
+	}
+
+	// Check the sender is the owner of the post
+	if post.Author != msg.Owner {
+		return nil, errors.Wrapf(sdkerrors.ErrInvalidRequest, "you are not the owner of the post")
+	}
+
+	// Check the permission to move the post
+	if !k.HasPermission(ctx, msg.TargetSubspaceID, msg.TargetSectionID, msg.Owner, types.PermissionWrite) {
+		return nil, errors.Wrap(subspacestypes.ErrPermissionDenied, "you don't have write permission on the target section")
+	}
+
+	// Get the next post id
+	newPostID, err := k.GetNextPostID(ctx, msg.TargetSubspaceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Update the post and validate it
+	updateTime := ctx.BlockTime()
+	postMove := types.NewPostMove(msg.TargetSubspaceID, msg.TargetSectionID, newPostID, updateTime)
+	updatedPost := postMove.Update(post)
+	err = k.ValidatePost(ctx, updatedPost)
+	if err != nil {
+		return nil, err
+	}
+
+	k.Keeper.SavePost(ctx, updatedPost)
+
+	// Update the id for the next post
+	k.SetNextPostID(ctx, msg.TargetSubspaceID, newPostID+1)
+
+	// Move all post attachments
+	newAttachmentID := uint32(0)
+	k.IteratePostAttachments(ctx, msg.SubspaceID, msg.PostID, func(attachment types.Attachment) (stop bool) {
+		newAttachmentID++
+		attachmentMove := types.NewAttachmentMove(msg.TargetSubspaceID, newPostID, newAttachmentID)
+		updatedAttachment := attachmentMove.Update(attachment)
+		k.SaveAttachment(ctx, updatedAttachment)
+
+		if types.IsActivePoll(updatedAttachment) {
+			k.InsertActivePollQueue(ctx, updatedAttachment)
+		}
+		return false
+	})
```

Comment:
Why is this being done? Why can't we preserve the attachment ids that were used inside the old post instead? They are unique to the post itself anyway. If we preserve them we have less writes (which means less gas used) and we can simply set the `NextAttachmentID` to be `max(post attachment ids) + 1` which is way simpler. This way we would also make it easier for external tools that could simply update the post id instead of having to delete the attachments and re-storing them as well

Reply:
Agree with you, we should preserve the original attachment id is correct, and if anything wrong invariant will alert when simulation test. I was confused that we should reset id or apply the original id here, finally I decided to reset the id because it will be error if ids are wrongly set.

By the way, I don't think it brings less writes since it actually adds some keys for moved attachments then deleted the original ones. The storage side is performing the same operations like:
```go
store.Set(movedAttachmentKey, movedAttachment)
store.Delete(oldAttachmentKey, oldAttachment)
```